src/module.rs

```diff
@@ -30,41 +30,96 @@ impl WasmModule {
+                     // the field of External::Function(_) is an index into
+                     // the type section.
+                     let name = import.field();
+                     let ty = self.get_type(*tyid)
+                         .expect(format!("Couldn't get type {} for imported function {}",
+                                         tyid,
+                                         name)
+                                         .as_str());
+
+                     Some(WasmFunction {
+                              // id is the index in the function index space.
+                              // An imported function's id is its order in the import section.
+                              id: i,
+                              ty,
+                              name: Some(name),
+                              body: None,
+                              source: SourceSection::Import,
+                          })
+                 } else {
+                     None
+                 })
+    }
+
+    /// Iterates over the function index space of the module.
+    /// According to the [WebAssembly design docs](https://github.com/sunfishcode/wasm-reference-manual/blob/master/WebAssembly.md):
+    ///
+    /// > The function index space begins with an index for each imported
+    /// > function, in the order the imports appear in the Import Section,
+    /// > if present, followed by an index for each function in the Function Section,
+    /// > if present, in the order of that section.
     pub fn functions(&self) -> impl Iterator<Item = WasmFunction> {
         let function_count = self.module.functions_space();
         if function_count == 0 {
             return Either::Left(iter::empty::<WasmFunction>());
         }
 
-        let bodies = self.function_bodies();
-        let types = self.function_types();
-        assert_eq!(function_count, bodies.size_hint().0);
-        assert_eq!(function_count, types.size_hint().0);
+        let imported_functions = self.imported_functions();
+        let imported_count = self.count_imported_functions();
+        let own_count = self.count_own_functions();
+        assert_eq!(function_count, imported_count + own_count);
 
-        let functions = types
-            .zip(bodies)
+        let own_functions = self.function_types()
+            .zip(self.function_bodies())
             .enumerate()
-            .map(move |(id, (ty, body))| {
-                     let name = self.get_function_name(id);
-                     let is_export = name.is_some(); // TODO(slim): Make this more robust later.
-                     WasmFunction::new(id, ty, name, body, is_export)
-                 });
-
-        Either::Right(functions)
-    }
+            .map(move |(i, (ty, body))| {
+                // Functions from the module function section appear
+                // after imported functions, in the index space.
+                let id = imported_count + i;
```

Comment:
So this is actually wrong since [all types of imports are stored in memory](https://github.com/sunfishcode/wasm-reference-manual/blob/master/WebAssembly.md#import-section). To start at the right index for functions, we need to use `import_count(&self, ImportCountType)` with all `ImportCountType`s.

Reply:
Per the [WebAssembly reference](https://github.com/sunfishcode/wasm-reference-manual/blob/master/WebAssembly.md#function-index-space):

> The *function index space* begins with an index for each imported function, in the order the imports appear in the [Import Section](https://github.com/sunfishcode/wasm-reference-manual/blob/master/WebAssembly.md#import-section), if present, followed by an index for each function in the [Function Section](https://github.com/sunfishcode/wasm-reference-manual/blob/master/WebAssembly.md#function-section), if present, in the order of that section.